### PR TITLE
perf(dispersion): optimize DFT-D3 kernels

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Optimize DFT-D3 Warp kernels for matrix and neighbor-list paths, split virial and non-virial launch paths, and update the JAX binding to dispatch kernels by position precision so float64 geometry is preserved through distance and periodic-shift calculations.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

None.

## Changes Made

- Add per-atom tiled launches and explicit block strides for DFT-D3 coordination, direct-force, CN-force, and Cartesian-shift kernels.
- Split DFT-D3 direct-force and CN-force kernels into virial and non-virial overloads for both matrix and neighbor-list formats.
- Update JAX DFT-D3 wrappers to dispatch Warp kernels by float32 or float64 position dtype while keeping D3 parameters and outputs in float32.
- Add DFT-D3 benchmark options for neighbor format and virial computation.
- Add JAX DFT-D3 tests for float64 large-origin displacement and periodic shift cancellation cases.

## Testing

- [ ] Unit tests pass locally (`make pytest`) - Not run.
- [ ] Linting passes (`make lint`) - Not run.
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

None.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Comment by dallasfoster** [comment]
/ok to test [25b2379](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/81/commits/25b2379f3849efd99145cbcf16ffa83e92b3725c)

**Comment by laserkelvin** [comment]
/ok to test [e370e57](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/81/commits/e370e57ff8dc57f024cbb516a3046ca6afcf92c3)

File: nvalchemiops/jax/interactions/dispersion/_dftd3.py [hunk 1]
```diff
@@ -88,63 +94,141 @@
...
-    wp_cn_forces_contrib_nm[wp.float32],
+_cn_forces_contrib_nm = _make_jax_kernels(
+    wp_cn_forces_contrib_nm,
+    num_outputs=1,
+    in_out_argnames=["forces"],
+)
+_cn_forces_contrib_nm_virial = _make_jax_kernels(
+    wp_cn_forces_contrib_nm_virial,
     num_outputs=2,
     in_out_argnames=["forces", "virial"],
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/jax/interactions/dispersion/_dftd3.py:410]: Fixed in e9fa08956ab3c1b6f2ecb8f3827940f717288464.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py [hunk 2]
```diff
@@ -645,46 +646,92 @@ def run_dftd3_nvalchemiops_benchmark(
...
-                k1=dftd3_config["k1"],
-                k3=dftd3_config["k3"],
-                batch_idx=batch_idx,
-                s5_smoothing_on=dftd3_config["s5_smoothing_on"],
-                s5_smoothing_off=dftd3_config["s5_smoothing_off"],
-                device=device,
+        if return_neighbor_list:
+            neighbor_list_data = system_data["neighbor_data"]  # (2, num_pairs)
+            neighbor_ptr = system_data["num_neighbor_data"]  # (N+1,)
+            unit_shifts = system_data["unit_shifts"] if compute_virial else None
```
Comment (laserkelvin) [inline benchmarks/interactions/dispersion/benchmark_dftd3.py]: This drops PBC shifts whenever `compute_virial` is false. The neighbor list was still built with PBC, so energy and force distances can be wrong. Pass shifts and cell whenever they exist; use `compute_virial` only for virial output.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py [hunk 3]
```diff
@@ -645,46 +646,92 @@ def run_dftd3_nvalchemiops_benchmark(
...
+
+            def dftd3_call():
+                return torch_dftd3(
+                    positions=positions,
+                    numbers=numbers,
+                    d3_params=d3_params,
+                    neighbor_list=neighbor_list_data,
+                    neighbor_ptr=neighbor_ptr,
+                    unit_shifts=unit_shifts,
+                    cell=cell if compute_virial else None,
```
Comment (laserkelvin) [inline benchmarks/interactions/dispersion/benchmark_dftd3.py]: This cell is still in Angstrom while `positions` are in Bohr. Shifted PBC distances and virials will be scaled wrong. Convert the cell to Bohr before passing it, and keep single-system torch cells shaped as `(1, 3, 3)`.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py [hunk 4]
```diff
@@ -1103,6 +1150,18 @@ def main():
         type=str,
         help="Override GPU SKU name for output files (default: auto-detect)",
     )
+    parser.add_argument(
```
Comment (laserkelvin) [inline benchmarks/interactions/dispersion/benchmark_dftd3.py:1163]: These flags only affect the torch path, and the CSV rows and filenames do not record them. Please reject them for unsupported backends or include the selected mode in every result and output name.

File: nvalchemiops/interactions/dispersion/_dftd3.py [hunk 5]
```diff
@@ -240,13 +240,30 @@
...
     "_cn_kernel_overload",
     "_direct_forces_and_dE_dCN_kernel_overload",
+    "_direct_forces_and_dE_dCN_kernel_virial_overload",
     "_cn_forces_contrib_kernel_overload",
+    "_cn_forces_contrib_kernel_virial_overload",
 ]
 
+# Threads per atom for _direct_forces_and_dE_dCN_kernel_matrix (matrix and NL).
+# 1 warp/atom; reductions via warp shuffles (no shared memory).
+DFTD3_MATRIX_DIRECT_FORCES_BLOCK_SIZE = 32
```
Comment (laserkelvin) [inline nvalchemiops/interactions/dispersion/_dftd3.py:256]: The new block-stride kernels need a test with more neighbors than the block size. Small dimer tests will not exercise the `neighbor_idx += block_stride` or `edge_idx += block_stride` paths.

File: nvalchemiops/jax/interactions/dispersion/_dftd3.py [hunk 6]
```diff
@@ -88,63 +94,427 @@
...
+    wp_direct_forces_kernel_nl,
+    num_outputs=3,
+    in_out_argnames=["dE_dCN", "forces", "energy"],
+)
+_direct_forces_kernel_nl_virial = _make_jax_kernels(
+    wp_direct_forces_kernel_nl_virial,
     num_outputs=4,
     in_out_argnames=["dE_dCN", "forces", "energy", "virial"],
 )
+direct_forces_kernel_nm_virial = _direct_forces_kernel_nm_virial[jnp.float32]
```
Comment (laserkelvin) [inline nvalchemiops/jax/interactions/dispersion/_dftd3.py]: These exported virial wrappers are fixed to float32, unlike the new dtype-dispatched private wrappers. Please dispatch them by dtype too, or keep only the compatibility wrappers public.

File: nvalchemiops/jax/interactions/dispersion/_dftd3.py [hunk 7]
```diff
@@ -88,63 +94,427 @@
...
+    compute_virial: bool,
+    dE_dCN: jax.Array,
+    forces: jax.Array,
+    energy: jax.Array,
+    virial: jax.Array,
+    **kwargs,
+) -> tuple[jax.Array, jax.Array, jax.Array, jax.Array]:
+    """Compatibility wrapper for the exported neighbor-matrix direct-force kernel."""
+    kernel_dtype = _normalize_dtype(positions.dtype)
+    kwargs.pop("launch_dims", None)
```
Comment (laserkelvin) [inline nvalchemiops/jax/interactions/dispersion/_dftd3.py]: This silently ignores caller-provided `launch_dims`. Please validate that it matches `(positions.shape[0], JAX_DFTD3_BLOCK_DIM)` and raise a clear error if it does not.

File: nvalchemiops/jax/interactions/dispersion/_dftd3.py [hunk 8]
```diff
@@ -584,113 +1030,174 @@ def _dftd3_nl_impl(
...
-        float(inv_w),
-        periodic,
-        batch_idx_i32,
-        compute_virial,
-        dE_dCN_init,
-        forces_init,
-        energy_init,
-        virial_init,
-    )
+    if compute_virial:
```
Comment (laserkelvin) [inline nvalchemiops/jax/interactions/dispersion/_dftd3.py:1138]: This new CSR virial branch needs a high-level test with nonzero `unit_shifts`, comparing CSR to the matrix path. Zero-shift non-virial tests will not catch shift conversion or virial bugs here.

File: test/interactions/dispersion/bindings/jax/test_dftd3.py [hunk 9]
```diff
@@ -409,6 +587,91 @@ def test_float64_positions(self, h2_system, functional_params, d3_params, device
         assert energy.dtype == jnp.float32
         assert coord_num.dtype == jnp.float32
 
+    def test_float64_positions_preserve_large_origin_displacement(
```
Comment (laserkelvin) [inline test/interactions/dispersion/bindings/jax/test_dftd3.py:787]: These float64 precision tests only cover matrix format. Please add or parametrize a CSR PBC case too, so the new neighbor-list dtype dispatch and cell casting are covered.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py [hunk 10]
```diff
@@ -645,46 +646,92 @@ def run_dftd3_nvalchemiops_benchmark(
...
+                    batch_idx=batch_idx,
+                    s5_smoothing_on=dftd3_config["s5_smoothing_on"],
+                    s5_smoothing_off=dftd3_config["s5_smoothing_off"],
+                    compute_virial=compute_virial,
+                    device=device,
+                )
+        else:
+            neighbor_matrix = system_data["neighbor_data"]
+            neighbor_matrix_shifts = (
+                system_data["unit_shifts"] if compute_virial else None
```
Comment (laserkelvin) [inline benchmarks/interactions/dispersion/benchmark_dftd3.py]: Same PBC issue in the matrix path: this drops shifts unless `compute_virial` is true. Energy and forces still need the shifts whenever the neighbor list was built with PBC.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py [hunk 11]
```diff
@@ -645,46 +646,92 @@ def run_dftd3_nvalchemiops_benchmark(
...
+
+            def dftd3_call():
+                return torch_dftd3(
+                    positions=positions,
+                    numbers=numbers,
+                    d3_params=d3_params,
+                    neighbor_matrix=neighbor_matrix,
+                    neighbor_matrix_shifts=neighbor_matrix_shifts,
+                    fill_value=total_atoms,
+                    cell=cell if compute_virial else None,
```
Comment (laserkelvin) [inline benchmarks/interactions/dispersion/benchmark_dftd3.py]: Same unit issue in the matrix path: this cell is still in Angstrom while `positions` are in Bohr. Convert the cell before passing it to DFT-D3.

**Comment by laserkelvin** [comment]
/ok to test [e9fa089](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/81/commits/e9fa08956ab3c1b6f2ecb8f3827940f717288464)

**Comment by laserkelvin** [comment]
/ok to test [3d0bb39](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/81/commits/3d0bb393cb998810b0effe2429788befdf4efceb)

File: benchmarks/interactions/dispersion/benchmark_dftd3.py (hunk 3 above)
Comment (zubatyuk) [reply benchmarks/interactions/dispersion/benchmark_dftd3.py]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py (hunk 10 above)
Comment (zubatyuk) [reply benchmarks/interactions/dispersion/benchmark_dftd3.py]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py (hunk 11 above)
Comment (zubatyuk) [reply benchmarks/interactions/dispersion/benchmark_dftd3.py]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py (hunk 4 above)
Comment (zubatyuk) [reply benchmarks/interactions/dispersion/benchmark_dftd3.py:1163]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: test/interactions/dispersion/bindings/jax/conftest.py [hunk 12]
```diff
@@ -19,4 +19,5 @@
 
 import pytest
 
-pytest.importorskip("jax", reason="No JAX installed.")
+jax = pytest.importorskip("jax", reason="No JAX installed.")
+jax.config.update("jax_enable_x64", True)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply test/interactions/dispersion/bindings/jax/conftest.py]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: test/interactions/dispersion/bindings/jax/test_dftd3.py (hunk 9 above)
Comment (zubatyuk) [reply test/interactions/dispersion/bindings/jax/test_dftd3.py:787]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: nvalchemiops/interactions/dispersion/_dftd3.py [hunk 13]
```diff
@@ -1204,35 +1447,37 @@ def _compute_cartesian_shifts(
 
     Notes
     -----
-    Launch with dim=num_atoms (one thread per atom). Each thread processes all edges
-    for that atom using the CSR pointers.
+    Launch with dim=(num_atoms, DFTD3_NL_CARTESIAN_SHIFTS_BLOCK_SIZE),
+    block_dim=DFTD3_NL_CARTESIAN_SHIFTS_BLOCK_SIZE. One block per atom; all threads
+    stride through the atom's CSR edge range with step=block_stride. Each thread
+    writes its own cartesian_shifts entries independently..
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/interactions/dispersion/_dftd3.py]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: benchmarks/interactions/dispersion/benchmark_dftd3.py (hunk 2 above)
Comment (zubatyuk) [reply benchmarks/interactions/dispersion/benchmark_dftd3.py]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: nvalchemiops/jax/interactions/dispersion/_dftd3.py (hunk 6 above)
Comment (zubatyuk) [reply nvalchemiops/jax/interactions/dispersion/_dftd3.py]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: nvalchemiops/jax/interactions/dispersion/_dftd3.py (hunk 8 above)
Comment (zubatyuk) [reply nvalchemiops/jax/interactions/dispersion/_dftd3.py:1138]: Fixed in 98f99b29c1f7afa6d883e0ebab47149b4a024caa.

File: test/interactions/dispersion/test_dftd3.py [hunk 14]
```diff
@@ -1187,6 +1189,174 @@ def test_nm_nl_consistency_hcl_dimer(self, request):
...
+        """Test matrix and CSR PBC paths with more neighbors than a block."""
+        device = request.getfixturevalue("device")
+        if "cuda" not in device:
+            pytest.skip("Block-stride coverage requires CUDA block launches")
+        element_tables = request.getfixturevalue("element_tables")
+        functional_params = request.getfixturevalue("functional_params")
+
+        num_atoms = 70
+        max_neighbors = num_atoms - 1
+        assert max_neighbors > 64
```
Comment (laserkelvin) [inline test/interactions/dispersion/test_dftd3.py]: This is pretty unnecessary?

**Review by laserkelvin — APPROVED** [review]
Just made the one comment, but otherwise LGTM

File: test/interactions/dispersion/test_dftd3.py (hunk 14 above)
Comment (zubatyuk) [reply test/interactions/dispersion/test_dftd3.py]: Fixed in fb6e549a2d07233b26962b646cb0411537aeea98.

**Comment by laserkelvin** [comment]
/ok to test [fb6e549](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/81/commits/fb6e549a2d07233b26962b646cb0411537aeea98)

**Review by laserkelvin — APPROVED** [review]
